File: python/LinkedList/day16/cloneALL.py
class Node:
    def __init__(self, data, next=None, random=None):
        self.data = data
        self.next = next
        self.random = random
    


def cloneALinkedList(head):
    if head is None:
        return None
    



    # mpp = {}
    # temp = head

    # while temp:
    #     newNode = Node(temp.data)
    #     mpp[temp] = newNode
    #     temp = temp.next
        
    
    # temp = head

    # while temp:
    #     copyNode = mpp[temp]
    #     copyNode.next = mpp.get(temp.next)
    #     copyNode.random = mpp.get(temp.random)
    #     temp = temp.next

    
    # return mpp[head]






    # optimal approach...

    # (1) Insert copy nodes in between...
    temp = head
    while temp:
        copyNode = Node(temp.data)
        copyNode.next = temp.next
        temp.next = copyNode
        temp = temp.next.next
    

    # (2) Connect the random pointer...
    temp = head

    while (temp):
        copyNode = temp.next
        
        if (temp.random):
            copyNode.random = temp.random.next
        else:
            copyNode.random = None
        
        temp = temp.next.next

    
    # (3) Connect the next pointer...
    dNode = Node(-1)
    res = dNode
    temp = head

    while temp:
        res.next = temp.next
        temp.next = temp.next.next
        res = res.next
        temp = temp.next
    
    return dNode.next







def printList(head):
    temp = head
    while temp:
        rand = temp.random.data if temp.random else -1
        print(f"{temp.data}({rand})", end=" -> ")
        temp = temp.next
    print("None")




# ------------ MAIN METHOD ------------
n = int(input("Enter number of nodes: "))
values = list(map(int, input("Enter node values: ").split()))

nodes = [Node(val) for val in values]

# Link next pointers...
for i in range(n - 1):
    nodes[i].next = nodes[i + 1]

print("Enter random pointer index for each node (-1 for null):")
rand_indices = list(map(int, input().split()))

for i in range(n):
    if rand_indices[i] != -1:
        nodes[i].random = nodes[rand_indices[i]]

head = nodes[0]

clonedHead = cloneALinkedList(head)

print("\nOriginal List:")
printList(head)

print("\nCloned List:")
printList(clonedHead)

File: java/LinkedList/Day16/CloneALL.java
package Day16;
import java.util.*;

class Node {
    int data;
    Node next;
    Node random;

    Node(int data1, Node next1, Node random1) {
        this.data = data1;
        this.next = next1;
        this.random = random1;
    }

    Node(int data1) {
        this.data = data1;
        this.next = null;
        this.random = null;
    }
}


public class CloneALL {
    private static Node cloneALL(Node head) {
        if (head == null) return null;

        Node temp = head;
        HashMap<Node, Node> mpp = new HashMap<>();

        while (temp != null) {
            Node newNode = new Node(temp.data);
            mpp.put(temp, newNode);
            temp = temp.next;
        }


        temp = head;

        while (temp != null) {
            Node copyNode = mpp.get(temp);
            copyNode.next = mpp.get(temp.next);
            copyNode.random = mpp.get(temp.random);
            temp = temp.next;
        }

        return mpp.get(head);
    }





    // Print linked list (data + random)...
    private static void printList(Node head) {
        Node temp = head;
        while (temp != null) {
            int randomVal = (temp.random != null) ? temp.random.data : -1;
            System.out.println(
                "Node data: " + temp.data + 
                ", Random points to: " + randomVal
            );
            temp = temp.next;
        }
    }

    // MAIN METHOD...
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of nodes: ");
        int n = sc.nextInt();

        Node[] nodes = new Node[n];

        System.out.print("Enter node values: ");
        for (int i = 0; i < n; i++) {
            nodes[i] = new Node(sc.nextInt());
        }

        // Link next pointers...
        for (int i = 0; i < n - 1; i++) {
            nodes[i].next = nodes[i + 1];
        }

        System.out.println(
            "Enter random pointer index for each node (-1 for null):"
        );
        for (int i = 0; i < n; i++) {
            int idx = sc.nextInt();
            if (idx != -1) {
                nodes[i].random = nodes[idx];
            }
        }

        Node head = nodes[0];

        Node clonedHead = cloneALL(head);

        System.out.println("\nOriginal List:");
        printList(head);

        System.out.println("\nCloned List:");
        printList(clonedHead);

        sc.close();
    }
}